### ISSUE TITLE
least_positive_float doesn't print/read correctly in float/bfloat

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2119]
Imported from SourceForge on 2024-07-05 18:43:05
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-02-24 02:47:08
Original: https://sourceforge.net/p/maxima/bugs/4107

---

printing the smallest positive float then reading that back in results in 0.0
~~~~
least_positive_float: 2.0^-1074$
print(least_positive_float)  => 4.940656458412465e-324
?print(least_positive_float) => 4.9406564584124654e-324
parse_string("4.940656458412465e-324") => 0.0
parse_string("4.9406564584124654e-324") => 4.9406564584124654e-324

parse_string(string(2.0b0^-1074))*2^1074 => 9.999...b-1
~~~~

Tested in both:
Maxima 5.46.0 SBCL 2.3.0 MacOS
Maxima 5.45.1 SBCL 2.0.0 Windows

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:07
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-02-24 15:48:40
Original: https://sourceforge.net/p/maxima/bugs/4107/#26b5

---

It's unfortunate, but I don't think maxima has ever had print/read consistency.

On the other hand, with cmucl:
```
parse_string("4.940656458412465e-324") => 4.940656458412465e-324
```
But I don't think Maxima has ever had print/read consistency because the float printer with default values wants to print numbers with 15 digits of precision, which isn't always enough.   It's all in `exploden-format-float` in commac.lisp.  I do not know why it does what it does.  I've always felt that if `fpprintprec` is zero, we should just use `cl:print` to get print/read consistency.  Well, except I didn't look to see how maxima reads numbers.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:10
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-02-24 16:04:42
Original: https://sourceforge.net/p/maxima/bugs/4107/#26b5/6f77

---

Yes, but these are really flagrant errors -- 100% error!
Correct rounding on read:
(0.0e0,2.4698e-324) => 0 
(2.4708e-324,7.4105e-324) => 4.9407e-324 
(7.4115e-324,1.2351e-323) => 9.8813e-324 
(1.2352e-323,1.7292e-323) => 1.4822e-323 
(1.7293e-323,2.2232e-323) => 1.9763e-323 
(2.2233e-323,2.7173e-323) => 2.4703e-323 
(2.7174e-323,3.2114e-323) => 2.9644e-323 
(3.2115e-323,3.7054e-323) => 3.4585e-323 
(3.7055e-323,4.1995e-323) => 3.9525e-323 
(4.1996e-323,4.6936e-323) => 4.4466e-323 
(4.6937e-323,5.1876e-323) => 4.9407e-323 

But Maxima gets the rounding grossly wrong: 4.9406e-323 => 4.4466e-323

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:14
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-24 17:57:52
Original: https://sourceforge.net/p/maxima/bugs/4107/#26b5/ea31

---

We can still implement print/read consistency even though Maxima hasn't had it yet; it's not too late.

The one topic that is undecided to me is how to handle `fpprintprec` other than default. If it's the default, we can punt to CL:PRINT. If it's not the default, how about this: in 2-d output, honor `fpprintprec`; if 1-d, honor an additional flag for print/read consistency and either ignore `fpprintprec` and implement print/read consistency, or honor `fpprintprec`. 

The reason I bring up a new flag for print/read consistency is that sometimes one still wants a certain number of digits in output, even if it's 1-d; it depends on the task. 

Punting to CL:PRINT means that the description of `fpprintprec` has to change. Mostly what I want in that respect is to minimize any Lisp implementation-specific stuff. Maybe the documentation can say that a float is printed with enough digits to reproduce the number on input when `fpprintprec` is 0.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:17
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-02-24 18:11:02
Original: https://sourceforge.net/p/maxima/bugs/4107/#26b5/ea31/9b4b

---

There seem to be two issues:
* Maxima READ doesn't round correctly for denormalized numbers (examples above). 
* Print/read consistency: agreed that fpprintprec=0 for this makes sense .

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:21
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-02-25 16:01:44
Original: https://sourceforge.net/p/maxima/bugs/4107/#c95a

---

I think it makes sense if `fpprintprec` is 0, that we use `cl:print`.  I don't think we should distinguish 1d and 2d output here.  That would be super confusing to get different numbers depending on whether it's 1d or 2d.

I tried `:lisp (read-from-string "4.9406e-323")` with both sbcl 2.0.1-8.fc36 and cmucl snapshot-2021-07 and the result in both cases is 4.446590812571219e-323.  This seems plausible though because `float(49406/10^(323+5))` returns 4.4465e-323.

One of your examples, Stavros involves bfloats.  I think it would be very difficult to get bfloats to have print/read consistency.  All of the float printers I've seen need to know the smallest (and largest) possible floating-point number and there just isn't any such value for bfloats.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:25
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-04-23 23:39:13
Original: https://sourceforge.net/p/maxima/bugs/4107/#4a87

---

Since the code for printing and parsing bigfloats is separate from ordinary floats, I've opened a separate bug report for that: https://sourceforge.net/p/maxima/bugs/4134/

The Maxima parser punts to the Lisp implementation to parse tokens via READ-FROM-STRING (see READLIST in src/nparse.lisp), so it would appear incorrect rounding on input is a Lisp implementation bug; maybe any such discrepancy should be reported to the respective project bug tracker.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:28
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-04-23 23:58:42
Original: https://sourceforge.net/p/maxima/bugs/4107/#91c1

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 18:43:32
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-04-23 23:58:42
Original: https://sourceforge.net/p/maxima/bugs/4107/#5f0b

---

I'm closing this report as fixed by commit [ f8d2e30 ]. 

Now floats are formatted by PRIN1-TO-STRING with `*PRINT-READABLY*` enabled when fpprintprec is 0 or greater than 16. (Otherwise floats are printed by FORMAT with `~f` or `~e` directive, depending on magnitude of the argument, as before.)

I've put in some tests to verify that some example values, and in particular `least_positive_float` and friends, are printed readably.

I tested the commit with ECL, CMUCL, Clisp, SBCL, and Clozure CL on Linux. I guess we'll see how it turns out with GCL (I can't build GCL 2.6.14 on my ancient development system).
